### PR TITLE
Sprites refactoring

### PR DESCRIPTION
**Pull request by Kubsy**
Sprites refactoring to be used Ideally with Instancing branch in TEN.

This branch adds DDS compression format and accompanies the changes made to sprites to TEN.

### CODE REVIEW

File: TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs [hunk 1]
```diff
@@ -219,17 +272,18 @@ void WriteTextureData(BinaryWriterEx writer)
             var sky = GetSkyTexture();
             using (var ms = new MemoryStream())
             {
-                using (var bmp = sky.ToBitmap())
-                    bmp.Save(ms, System.Drawing.Imaging.ImageFormat.Png);
                 writer.Write(sky.Width);
                 writer.Write(sky.Height);
-                var output = RemoveColorChunks(ms);
```
Comment (Lwmte) [inline TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs:226]: What was the purpose of `RemoveColorChunks` and why it was removed?

File: TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            int atlasWidth = 256;
+            int atlasHeight = 256;
+
+            do
+            {
+                RectPacker texPacker = new RectPackerTree(new VectorInt2(atlasWidth, atlasHeight));
+
+                done = true;
+
+                foreach (int UsedTextureIndex in usedTexturesProcessingOrder)
```
Comment (Lwmte) [inline TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs]: Why `UsedTextureIndex` is capitalized?

File: TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                    }
+                }
+            } while (!done);
+
+            ImageC resultingTexture = ImageC.CreateNew(atlasWidth, atlasHeight);
+
+            List<ImageC> resultingTextures = new List<ImageC>();
+            _usedTexturePackInfos = new Result[_usedTextures.Count];
+
+            for (int i=0;i<_usedTextures.Count;i++)
```
Comment (Lwmte) [inline TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs]: Please uncondense this line with spaces.

File: TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs [hunk 4]
```diff
@@ -63,14 +67,14 @@ private void BuildSprites()
             var spriteSequences = _level.Settings.WadGetAllSpriteSequences();
 
             // Add all sprites to the texture packer
-            var spriteAllocator = new Util.SpriteAllocator(_level.Settings.GameVersion == TRVersion.Game.TombEngine ? 2048 : 256);
+            var spriteAllocator = new TombEngineSpriteAllocator(4096);
```
Comment (Lwmte) [inline TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs]: Why there is a magic number of 4096? You should be able to get texture page size from `_limits` like this: `_limits[Limit.TexPageSize]`.

**Review by Lwmte — CHANGES_REQUESTED** [review]

File: TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs (hunk 2 above)
Comment (MontyTRC89) [reply TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs]: Copy and paste from somewhere. Fixed :-)

File: TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs (hunk 3 above)
Comment (MontyTRC89) [reply TombLib/TombLib/LevelData/Compilers/TombEngine/TombEngineSpriteAllocator.cs]: Fixed :-)

File: TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs (hunk 4 above)
Comment (MontyTRC89) [reply TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs]: Done!

File: TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs (hunk 1 above)
Comment (MontyTRC89) [reply TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs:226]: My mistake :-( it caused also the dark textures bug.

File: TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs [hunk 5]
```diff
@@ -89,33 +94,33 @@ private void BuildSprites()
...
                         continue;
                     }
 
-                    /*var packInfo = spriteAllocator.GetPackInfo(id);
-                 
-                    float x = packInfo.Pos.X / (float)_spritesTexturesPages[packInfo.OutputTextureID].Width;
-                    float y = packInfo.Pos.Y / (float)_spritesTexturesPages[packInfo.OutputTextureID].Height;
+                    var packInfo = spriteAllocator.GetPackInfo(id);
+
+                    int padding = 8;
```
Comment (Lwmte) [inline TombLib/TombLib/LevelData/Compilers/TombEngine/Textures.cs]: Why padding value is hardcoded? Can't we use padding specified in level settings?

**Review by Lwmte — APPROVED** [review]

**Pull request merged**